### PR TITLE
[WIP][NEW] Room file list

### LINKED ISSUE #34: [NEW] Files list
# Description

Add support to fetch the files list (paginated) in the SDK as a REST API.

## Progress

- [x] API request;
- [x] Paginated support;
- [x] Unit tests;

### PR DESCRIPTION
**Pull request by philipbrito**
This PR adds the method to get the file list from a room (via REST API).
Closes #34 

## Progress

- [x] API request -> get room file list and the total of files of a room.
     - [x] For Public channels.
     - [x] For Private Channels
     - [x] For One-to-One Channels
     - [x] For Livechat 
- [x] Pagination support.
- [x] Unit tests.

### CODE REVIEW

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 1]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
...
     }
 
+    void getRoomFiles(String roomId,
+                      BaseRoom.RoomType roomType,
+                      String offset,
+                      String sortBy,
+                      Sort sort,
+                      final RoomCallback.GetFilesCallback callback) {
+        checkNotNull(offset,"roomId == null");
+        checkNotNull(offset,"roomType == null");
```
Comment (rafaelks) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: `checkNotNull(roomType,"roomType == null");`

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 2]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
         });
     }
 
+    void getRoomFiles(String roomId,
+                      BaseRoom.RoomType roomType,
+                      String offset,
+                      String sortBy,
+                      Sort sort,
+                      final RoomCallback.GetFilesCallback callback) {
+        checkNotNull(offset,"roomId == null");
```
Comment (rafaelks) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: `checkNotNull(roomId,"roomId == null");`

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 3]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
...
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                if (!response.isSuccessful()) {
+                    processCallbackError(response, callback);
+                    return;
+                }
+
+                try {
+                    JSONObject json = new JSONObject(response.body().string());
+                    System.out.println("RESPONSE: " + json.toString());
```
Comment (rafaelks) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Do we have a way to only Log things in debugging mode?

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 4]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
...
+
+    void getRoomMembers() {
+
+    }
+
+    void getRoomFavoriteMessages() {
+
+    }
+
+    void getRoomPinnedMessages() {
```
Comment (rafaelks) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: What about separating the methods in different files? Isn't that better than doing everything in the same one?

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 5]
```diff
@@ -186,7 +271,7 @@ private HttpUrl requestUrl(HttpUrl baseUrl, String method) {
         return builder;
     }
 
-    private void procressCallbackError(Response response, Callback callback) {
+    private void processCallbackError(Response response, Callback callback) {
```
Comment (rafaelks) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java:304]: 👍

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 1 above)
Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: the good and old copy and paste and forget to edit

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 3 above)
Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: @filipedelimabrito use the `logger` for logging stuff...

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 4 above)
Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Maybe separate collections of methods into different files? Everything related to messages in one internal file, everything related to users in another.

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 5 above)
Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java:304]: Nice catch.

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 6]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
...
+                      final RoomCallback.GetFilesCallback callback) {
+        checkNotNull(offset,"roomId == null");
+        checkNotNull(offset,"roomType == null");
+        checkNotNull(offset,"offset == null");
+        checkNotNull(sortBy,"sortBy == null");
+        checkNotNull(sort,"sortDirection == null");
+        checkNotNull(callback,"callback == null");
+
+        String path;
+        switch (roomType) {
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Extract to a helper function, maybe two composed functions, one more generic, that will return only `channels.`, `groups.` or `dm.` and another composing the rest of the API path

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 7]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
...
+                    return;
+                }
+
+                try {
+                    JSONObject json = new JSONObject(response.body().string());
+                    System.out.println("RESPONSE: " + json.toString());
+
+                    JSONArray filesJSONArray = json.getJSONArray("files");
+                    int length = filesJSONArray.length();
+                    ArrayList<Attachment> attachments = new ArrayList<>();
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: This is not wrong, but the usual preferred method is to declare the variable only with the `List` interface, instead of its concrete type `ArrayList`, also, since you know the size of the array, you could initialize it with a proper initial size:
`List<Attachment> attachments = new ArrayList<>(lenght);`

I'm also planning to generify this handling code, using a generic response Class and moshi to parse it directly.

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 8]
```diff
@@ -163,6 +169,85 @@ public void onResponse(Call call, Response response) throws IOException {
...
+                    for (int i = 0; i < length; ++i) {
+                        attachments.add(new Attachment(filesJSONArray.getJSONObject(i)));
+                    }
+                    callback.onGetRoomFiles(json.optInt("total"), attachments);
+                } catch (JSONException e) {
+                    e.printStackTrace();
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: You should notify the callback about the error...

File: rocketchat-core/src/main/java/com/rocketchat/core/RocketChatClient.java [hunk 9]
```diff
@@ -144,6 +146,15 @@ public void pinMessage(String messageId, SimpleCallback callback) {
         restImpl.pinMessage(messageId, callback);
     }
 
+    public void getRoomFiles(String roomId,
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RocketChatClient.java]: lets try to keep a sane order on the methods.
Rest methods firts, and then ordered: at the top user creationg, signin, and server information (specially if it is unauthenticated), and then try to group related calls together.

File: rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java [hunk 10]
```diff
@@ -18,4 +19,7 @@
         void onGetRoomMembers(Integer total, List<User> members);
     }
 
-}
+    public interface GetFilesCallback extends Callback {
+        void onGetRoomFiles(Integer total, List<Attachment> files);
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java]: Maybe this could be just an `int` and avoid an autoboxing (I know that there are other parts using `Integer` already)

**Review by luciofm — CHANGES_REQUESTED** [review]

File: rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java (hunk 10 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java]: In the first moment I was surprised about the Integer usage, but I was following what was done. Of course we'll change it to int! Thanks for your thoughts!

File: rocketchat-core/src/main/java/com/rocketchat/core/RocketChatClient.java (hunk 9 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/RocketChatClient.java]: I agree with you.

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 7 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Yes! You are right! I forgot about this simple, but efficient details!

File: rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java (hunk 10 above)
Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java]: Actually we don't really need this, do we? this is total total or size of the files list (`files.size()`)?

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 11]
```diff
@@ -169,40 +178,42 @@ public void onResponse(Call call, Response response) throws IOException {
...
-            default:
-                path = "dm.files";
-                break;
-        }
-
-        RequestBody body = new FormBody.Builder()
-                .add("roomId", roomId)
-                .add("offset", offset)
-                .add("sort", "{\"" + sortBy + "\":" + sort.getDirection() + "}")
+        HttpUrl httpUrl = requestUrl(baseUrl, getRestApiMethodNameByRoomType(roomType) + "files")
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Maybe passing `"files"` as a parameter to  `getRestApiMethodNameByRoomType`, I think it will be cleaner

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 1 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Thanks for catching it @rafaelks and yeah what I was to said @luciofm 
Fortunately we will have tests!

File: rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java (hunk 10 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java]: A good usage of the `total` is on the current Rocket.Chat+ Android app. The total is a key that comes from the call that shows the total of files on a Room. This is total total

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 11 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Much cleaner! How we should named the second parameter? Is "method" good?

```
private String getRestApiMethodNameByRoomType(BaseRoom.RoomType roomType, String method) {
[...]
}
```

Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: I think that "method" is the full name (e.g. channels.files)... What do you think?

Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: I think method is ok...

File: rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java [hunk 12]
```diff
@@ -59,6 +61,13 @@ public void getMembers(RoomCallback.GetMembersCallback callback) {
         client.getRoomMembers(room.roomId(), false, callback);
     }
 
+    public void getFiles(String offset,
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java]: @filipedelimabrito please add a java doc for this method, since this is the entry point for this method.

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 13]
```diff
@@ -37,12 +44,60 @@
...
 
     RestImpl(OkHttpClient client, Moshi moshi, HttpUrl baseUrl, TokenProvider tokenProvider, Logger logger) {
         this.client = client;
         this.moshi = moshi;
         this.baseUrl = baseUrl;
         this.tokenProvider = tokenProvider;
+        this.logger = logger;
+    }
+
+    void signin(String username, String password, final LoginCallback loginCallback) {
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java:58]: Next time, if you are going to reorder methods (which is a good thing, don't get me wrong), do in a separate PR just for that... New code mixed with code reordering makes it hard to review :smiley:

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java [hunk 14]
```diff
@@ -148,34 +225,71 @@ public void onFailure(Call call, IOException e) {
...
+                return "dm." + method;
+        }
+    }
+
+    /**
+     * Builds the request URL as {baseUrl}/api/v1/{method}
+     *
+     * @param baseUrl The base URL.
+     * @param method The method name.
+     * @return A HttpUrl pointing to the REST API call.
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: please update the return type, it now returns an `HttpUrl.Builder`

Comment (luciofm) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java]: Anyway, this is a private method, let me check how we can filter it from javadocs generation

File: rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java [hunk 15]
```diff
@@ -18,4 +19,7 @@
         void onGetRoomMembers(Integer total, List<User> members);
     }
 
-}
+    public interface GetFilesCallback extends Callback {
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java]: take a look on `PaginatedCallback<T>` it is not being used yet, so you can tweak it for this...

File: rocketchat-core/src/main/java/com/rocketchat/core/model/attachment/Attachment.java [hunk 16]
```diff
@@ -8,26 +10,32 @@
 import org.json.JSONObject;
 
 /**
- * Created by sachin on 20/8/17.
+ * @author Sachin Shinde
+ * @author Filipe de Lima Brito (filipedelimabrito@gmail.com)
  */
-
 public class Attachment {
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/model/attachment/Attachment.java:16]: Could you try to convert this to `AutoValue`?  You can look on the current app.

Don't forget to add a method like this
```java
    public static JsonAdapter<Attachment> jsonAdapter(Moshi moshi) {
        return new AutoValue_Attachment.MoshiJsonAdapter(moshi);
    }
```
so we can generate the Json Parser for it.

File: sample/src/main/java/com/rocketchat/sample/RestApiRoomList.java [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+
+import java.util.List;
+import java.util.Scanner;
+
+/**
+ * This class shows the lists fetched from a Room (e.g. its member list, file list, favorite message list and pinned message list).
+ *
+ * @author Filipe de Lima Brito (filipedelimabrito@gmail.com)
+ */
+public class RestApiRoomList {
```
Comment (luciofm) [inline sample/src/main/java/com/rocketchat/sample/RestApiRoomList.java:27]: As discussed yesterday, just remove this file and provide just the snippet on `getRoomFilesByRoom()` method on the javadoc.

**Review by luciofm — CHANGES_REQUESTED** [review]

File: sample/src/main/java/com/rocketchat/sample/RestApiRoomList.java (hunk 17 above)
Comment (philipbrito) [reply sample/src/main/java/com/rocketchat/sample/RestApiRoomList.java:27]: Can we remove the entire sample module?

File: rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java (hunk 15 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/callback/RoomCallback.java]: Thanks for advice!

File: rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java (hunk 13 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/RestImpl.java:58]: Ohh, sorry! I'll pay attention on it next time!

File: rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java (hunk 12 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java]: Added!

File: sample/src/main/java/com/rocketchat/sample/RestApiRoomList.java (hunk 17 above)
Comment (luciofm) [reply sample/src/main/java/com/rocketchat/sample/RestApiRoomList.java:27]: @filipedelimabrito I would make a separate PR for that... Let's keep PRs focused.

File: rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java [hunk 18]
```diff
@@ -65,14 +66,28 @@ public void getMembers(RoomCallback.GetMembersCallback callback) {
...
      * @param sortBy The attribute name to sort.
      * @param sort The order in which the results should be returned.
      * @param callback The paginated callback.
-     * @since 0.8.0
      * @see BaseRoom
      * @see com.rocketchat.core.model.attachment.Attachment.SortBy
      * @see Sort
+     * @since 0.8.0
      */
     public void getFiles(String offset,
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java]: Why the offset is a String?

Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java]: This is the way that it is designated on the API (our REST API). We could use an `int` and do a conversion on the final process of our SDK, but I think that it is unnecessary.

File: rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java [hunk 19]
```diff
@@ -59,6 +63,39 @@ public void getMembers(RoomCallback.GetMembersCallback callback) {
         client.getRoomMembers(room.roomId(), false, callback);
     }
 
+    /**
+     * Gets the room file list.
+     *
+     * <p>Example of expected usage:
+     *
+     * <blockquote><pre>
+     * room.getFiles("0", Attachment.SortBy.UPLOADED_DATE, Sort.DESC, new PaginatedCallback() {
```
Comment (luciofm) [inline rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java]: Oops, forgot to change the sample too

File: rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java (hunk 18 above)
Comment (philipbrito) [reply rocketchat-core/src/main/java/com/rocketchat/core/ChatRoom.java]: @luciofm I changed the `offset` to int. Now we are using the `String.valueOf()` to work with our call.

File: rocketchat-core/src/test/java/com/rocketchat/core/RestImplTest.java [hunk 20]
```diff
@@ -88,51 +116,115 @@ public void tesSigninShouldBeSuccessfull() {
...
+
+        RocketChatException exception = exceptionCaptor.getValue();
+        assertThat(exception.getMessage(), is(equalTo("A JSONObject text must begin with '{' at character 0")));
+        assertThat(exception.getCause(), is(instanceOf(JSONException.class)));
+    }
+
+    @Test
+    //TODO Needs to know why it is failing.
+    public void testGetRoomFilesShouldBeSuccessful() {
+        rest.getRoomFiles("general", BaseRoom.RoomType.PUBLIC, 0, Attachment.SortBy.UPLOADED_DATE, Sort.DESC, paginatedCallback);
```
Comment (luciofm) [inline rocketchat-core/src/test/java/com/rocketchat/core/RestImplTest.java]: You need  to configure the `mockServer` to expect a `files` call and return a response...

Comment (luciofm) [reply rocketchat-core/src/test/java/com/rocketchat/core/RestImplTest.java]: Check `tesSigninShouldBeSuccessful()` for that

**Review by luciofm — CHANGES_REQUESTED** [review]
Almost there

**Review by luciofm — APPROVED** [review]
